File: source/private/simpleLevenshtein.m
function distance = simpleLevenshtein(first, second)
%SIMPLELEVENSHTEIN Simplified function to compute the distance between two
%variants based on the Levenshtein distance.

% Copyright 2023 Acadèmia de su Sardu APS

first = iReplaceDiacritics(first);
second = iReplaceDiacritics(second);
distance = levenshteinCore(first, second);
end


function replaced = iReplaceDiacritics(inString)
% Replace all the characters with diacritic-less characters as specified in
% the the replacement table.

replaced = inString;
[before, after] = iReplacementTable();
for k = 1:numel(before)
    substituteTo = after{k};
    listOfCharctersToSubstitute = before{k};

    for c = 1:length(listOfCharctersToSubstitute)
        currCharacterToSubstitute = listOfCharctersToSubstitute(c);
        replaced = replace(replaced, currCharacterToSubstitute, substituteTo);
    end
end
end


function [before, after] = iReplacementTable()
conversionData = ...
    {'a', 'àá' ; ...
    'e', 'èé' ; ...
    'i', 'ìíï'; ...
    'o', 'òó' ; ...
    'u', 'ùú' ; ...
    'A', 'ÀÁ' ; ...
    'E', 'ÈÉ' ; ...
    'I', 'ÌÍ' ; ...
    'O', 'ÒÓ' ; ...
    'U', 'ÙÚ'};
before = conversionData(:, 2);
after = conversionData(:, 1);
end

File: python/linguistic_variation_toolbox/_private/simple_levenshtein.py
from .levenshtein_core import levenshtein_core

_REPLACEMENT_TABLE = [
    ("àá", "a"),
    ("èé", "e"),
    ("ìíï", "i"),
    ("òó", "o"),
    ("ùú", "u"),
    ("ÀÁ", "A"),
    ("ÈÉ", "E"),
    ("ÌÍ", "I"),
    ("ÒÓ", "O"),
    ("ÙÚ", "U"),
    ("-", ""),
]


def _replace_diacritics(s: str) -> str:
    for chars, replacement in _REPLACEMENT_TABLE:
        for ch in chars:
            s = s.replace(ch, replacement)
    return s


def simple_levenshtein(first: str, second: str) -> int:
    return levenshtein_core(_replace_diacritics(first), _replace_diacritics(second))
